### PR TITLE
chore: add DevContainer config for GitHub codespaces

### PR DESCRIPTION
**Pull request by kumaraditya303**
## Motivation

Dev container for GitHub codespaces
### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes
## Test Plan
Not Required
## Related PRs
None

### CODE REVIEW

**Comment by slorber** [comment]
Thanks

Can you give context about this feature?

Why is it useful, and how can I validate that your PR is good?
Please don't assume that I know anything about Github codespaces

**Comment by kumaraditya303** [comment]
> Thanks
> 
> Can you give context about this feature?
> 
> Why is it useful, and how can I validate that your PR is good?
> Please don't assume that I know anything about Github codespaces

@slorber Github codespaces uses docker containers as a complete dev environment and it can be configured as per the project needs. This pr add the configuration required for docusaurus i.e. install all the packages and forwards the docusaurus port for the user to connect using browsers otherwise users cannot open it.

**Comment by slorber** [comment]
@kumaraditya303 thanks but I need more infos.

How can I test your PR exactly and see in practice that it works, and the benefits of supporting Codespaces?

Also, how do we document this so that users can discover this feature?

**Comment by kumaraditya303** [comment]
> @kumaraditya303 thanks but I need more infos.
> 
> How can I test your PR exactly and see in practice that it works, and the benefits of supporting Codespaces?
> 
> Also, how do we document this so that users can discover this feature?

@slorber It similar to gitpod and there isn't any docs needed as GitHub codespaces automatically uses devcontainer.json. Users can directly open with codespaces and good to go.

**Comment by kumaraditya303** [comment]
@slorber do you have access to GitHub codespaces?

**Comment by slorber** [comment]
Unfortunately no, I'm still on the waiting list 😅

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
+	"name": "Docusaurus Dev Container",
+	"image": "mcr.microsoft.com/vscode/devcontainers/typescript-node:0-10-buster",
```
Comment (slorber) [inline .devcontainer/devcontainer.json:3]: how did you choose the image here?

**Comment by kumaraditya303** [comment]
Ohh, I am a beta and pro user so I got it very early.

File: .devcontainer/devcontainer.json (hunk 1 above)
Comment (slorber) [reply .devcontainer/devcontainer.json:3]: We may need something on node LTS 14 instead of node 10

**Comment by slorber** [comment]
Can you show me screenshots or somthing about how this feature works?

I'm ok to merge it as it seems harmful and could check later how it works, just curious to see it in action

File: .devcontainer/devcontainer.json (hunk 1 above)
Comment (kumaraditya303) [reply .devcontainer/devcontainer.json:3]: Because the GitHub workflow are using nodejs 10 so I chose that

Comment (kumaraditya303) [reply .devcontainer/devcontainer.json:3]: The image I chose is one of the official GitHub codespaces images for nodejs with typescript support

Comment (slorber) [reply .devcontainer/devcontainer.json:3]: We may move/require node 14 LTS soon so it's better to use node 14 directly for this feature

File: .devcontainer/devcontainer.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{
+	"name": "Docusaurus Dev Container",
```
Comment (RDIL) [inline .devcontainer/devcontainer.json:2]: Why tabs?

**Review by RDIL — CHANGES_REQUESTED** [review]

**Comment by RDIL** [comment]
> > @kumaraditya303 thanks but I need more infos.
> > How can I test your PR exactly and see in practice that it works, and the benefits of supporting Codespaces?
> > Also, how do we document this so that users can discover this feature?
> 
> @slorber It similar to gitpod and there isn't any docs needed as GitHub codespaces automatically uses devcontainer.json. Users can directly open with codespaces and good to go.

IMO its a ripoff of Gitpod that is less extendable and I don't see any benefit of using it over Gitpod.  Gitpod, unlike Codespaces, is fully open-source.

**Comment by kumaraditya303** [comment]
@slorber 
Untitled

**Comment by kumaraditya303** [comment]
> Can you show me screenshots or somthing about how this feature works?
> 
> I'm ok to merge it as it seems harmful and could check later how it works, just curious to see it in action

@slorber  What do you mean by `harmful` ?

**Comment by slorber** [comment]
I meant "does not seem harmful".

Let's merge it and see if people report config problems.
Maybe Gitpod is better, maybe not, it's probably not a bad idea to support both anyway

File: .devcontainer/devcontainer.json (hunk 2 above)
Comment (slorber) [reply .devcontainer/devcontainer.json:2]: just reformatted it

**Comment by slorber** [comment]
@kumaraditya303 can you reformat the json file with prettier please?

Or give me edit permission on your PR so that I can do it myself? (otherwise I'd need to open another PR)

**Comment by kumaraditya303** [comment]
@slorber given edit permission

**Comment by slorber** [comment]
thanks

**Pull request merged**